ast/src/analyzed/types.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt::Display;
+
+use powdr_number::FieldElement;
+
+use crate::parsed::{ArrayTypeName, Expression, FunctionTypeName, TupleTypeName, TypeName};
+
+use super::Reference;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct TypedExpression<T, Ref = Reference> {
+    pub e: Expression<T, Ref>,
+    pub ty: Option<Type>,
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum Type {
+    /// Boolean
+    Bool,
+    /// Integer (arbitrary precision)
+    Int,
+    /// Field element (unspecified field)
+    Fe,
```

Comment:
eventually we could also support specifying the field in the type, like `(BV 256), (BV 32)` in smtlib2

Reply:
True, but I kind of like this, since we want to be generic. Of course, it could be `Fe<n>` as well or even `Fe<n: n >= 2**32>`.

Reply:
it could be a const generic, but yea future stuff